### ISSUE TITLE
Allow `MenuAnchor` created menus to traverse the submenu via keyboard.

### ISSUE THREAD

**Issue by gspencergoog** [flutter/flutter#119532]
Currently, if you create a menu using `MenuAnchor`, and you tab to the button, you can activate it, but you can't traverse to the submenu using the keyboard.  This is because the submenu is in the overlay, and the button isn't handling the key events.

I'm not quite sure what the fix for this is, because it may be that the button itself has to add a shortcut for the arrow and tab keys to get to the submenu. Or maybe it just needs to focus the first item in the submenu when activated.

**Comment by huycozy** [comment]
Hi @gspencergoog,
There is a reproduction case at https://github.com/flutter/flutter/issues/144381 which pointed out that only `Tab` key doesn't work on `SubmenuButton`. I see `Arrow` keys still works. I wonder if this issue assemble that case?

**Comment by davidhicks980** [comment]
So I actually think this issue is distinct from https://github.com/flutter/flutter/issues/144381, but is the same as another issue I filed here: https://github.com/flutter/flutter/issues/147770.

**Comment by PvtPuddles** [comment]
I've just bumped into this issue as well (`Flutter 3.22.3 • channel stable • Framework • revision b0850beeb2 (3 weeks ago) • 2024-07-16 21:43:41 -0700`).

I have a column of menu anchors.  When you press the up arrow (when the menu is above the anchor), you expect to enter the menu, or you expect to already have the menu in focus (as Spencer mentioned).  Instead, the focus moves up to the previous menu anchor in the list.

> There is a reproduction case at #144381 which pointed out that only `Tab` key doesn't work on `SubmenuButton`. I see `Arrow` keys still works. I wonder if this issue assemble that case?

When tabbing into the anchor and then activating the menu, tab no longer performs any action.  I can still use the arrows to navigate the column of anchors , though "arrow-ing" off the anchor leaves its menu open.

Weirdly, I found that by setting the `MenuAnchor.childFocusNode` I can consistently get into the menu by pressing the right arrow (and only the right arrow), which selects the first menu item.

**Comment by davidhicks980** [comment]
@PvtPuddles Can you post an example and/or try it on master? I think some of the focus changes I PR'd have landed on master.

**Comment by PvtPuddles** [comment]
@davidhicks980 You accidentally sent me down quite the rabbit hole.

As it stands (`Flutter 3.24.0-1.0.pre.505 • channel master • Framework • revision 64b373b1f2 (15 minutes ago) • 2024-08-08 01:39:24 +0200`):
1. tab does not work (as Huy mentioned)
2. the arrow keys *do* work as expected, so long as the `MenuAnchor.menuStyle` is not padded.
3. Losing focus of the `MenuAnchor` does not close its menu still (this is tangential, but I figured I'd mention it as I haven't found a way to do this manually)

Per 2, the padding seems to prevent the first item from receiving focus, as in the example below.  Interestingly, the second item will receive focus next (unlike my production app, in which focus never makes it into the menu). 

Here is the code I've ripped out of our project, hopefully you can reproduce:

```dart
import 'package:flutter/material.dart';

class ExampleApp extends StatelessWidget {
  const ExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    final content = Column(
      children: [
        for (int i = 0; i < 3; i++)
          _DropdownMenu(
            menuItems: [
              for (int j = 0; j < 3; j++)
                MenuItemButton(
                  child: Text("$i - $j"),
                  onPressed: () {},
                ),
            ],
            trailingIcon: IconButton(
              onPressed: () {},
              icon: const Icon(Icons.clear),
            ),
            child: Text("Dropdown menu $i"),
          )
      ],
    );

    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text("Focus Experiments"),
        ),
        body: SafeArea(
          minimum: const EdgeInsets.all(16),
          child: content,
        ),
      ),
    );
  }
}

/// A custom implementation bridging the gap between a [DropdownMenu] and a
/// [DropdownButton]
class _DropdownMenu extends StatefulWidget {
  const _DropdownMenu({
    required this.menuItems,
    this.trailingIcon,
    required this.child,
  });

  final List<Widget> menuItems;
  final Widget? trailingIcon;
  final Widget child;

  @override
  State<_DropdownMenu> createState() => _DropdownMenuState();
}

class _DropdownMenuState extends State<_DropdownMenu> {
  final _controller = WidgetStatesController();
  final childNode = FocusNode(debugLabel: "Dropdown Inkwell");

  @override
  void dispose() {
    _controller.dispose();
    childNode.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MenuAnchor(
      childFocusNode: childNode,
      style: const MenuStyle(
        padding: WidgetStatePropertyAll(
            EdgeInsets.symmetric(vertical: 8, horizontal: 16)),
      ),
      builder: (context, controller, child) {
        return InkWell(
          focusNode: childNode,
          statesController: _controller,
          borderRadius: const BorderRadius.vertical(top: Radius.circular(8)),
          onTap: controller.open,
          child: ListenableBuilder(
            listenable: _controller,
            builder: (context, _) {
              return InputDecorator(
                decoration:
                    const InputDecoration(border: UnderlineInputBorder()),
                isFocused: _controller.value.contains(WidgetState.focused),
                isHovering: _controller.value.contains(WidgetState.hovered),
                child: Row(
                  children: [
                    Expanded(
                      child: DefaultTextStyle(
                          style:
                              Theme.of(context).dropdownMenuTheme.textStyle ??
                                  Theme.of(context).textTheme.bodyLarge!,
                          child: child!),
                    ),
                    if (widget.trailingIcon != null) widget.trailingIcon!,
                  ],
                ),
              );
            },
          ),
        );
      },
      menuChildren: widget.menuItems,
      child: widget.child,
    );
  }
}
```

**Comment by davidhicks980** [comment]
So the "correct" way to fix this will be coming in a future patch but requires some internal changes. That said, you can do some gymnastics... try this out (sorry about the spaghetti code -- this is just a quick fix and should be cleaned up):

```dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

void main(List<String> args) {
  runApp(const ExampleApp());
}

class ExampleApp extends StatefulWidget {
  const ExampleApp({super.key});

  @override
  State<ExampleApp> createState() => _ExampleAppState();
}

class _ExampleAppState extends State<ExampleApp> {
  List<FocusNode> firstItemFocusNodes =
      List.generate(3, (index) => FocusNode());

  @override
  void dispose() {
    for (final node in firstItemFocusNodes) {
      node.dispose();
    }

    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final content = Column(
      children: [
        for (int i = 0; i < 3; i++)
          _DropdownMenu(
            menuItems: [
              for (int j = 0; j < 3; j++)
                MenuItemButton(
                  focusNode: j == 0 ? firstItemFocusNodes[i] : null,
                  child: Text("$i - $j"),
                  onPressed: () {},
                ),
            ],
            trailingIcon: IconButton(
              onPressed: () {},
              icon: const Icon(Icons.clear),
            ),
            firstItemFocusNode: firstItemFocusNodes[i],
            child: Text("Dropdown menu $i"),
          ),
      ],
    );

    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text("Focus Experiments"),
        ),
        body: SafeArea(
          minimum: const EdgeInsets.all(16),
          child: content,
        ),
      ),
    );
  }
}

/// A custom implementation bridging the gap between a [DropdownMenu] and a
/// [DropdownButton]
class _DropdownMenu extends StatefulWidget {
  const _DropdownMenu({
    required this.menuItems,
    this.trailingIcon,
    required this.child,
    required this.firstItemFocusNode,
  });

  final List<Widget> menuItems;
  final Widget? trailingIcon;
  final Widget child;
  final FocusNode firstItemFocusNode;

  @override
  State<_DropdownMenu> createState() => _DropdownMenuState();
}

class _DropdownMenuState extends State<_DropdownMenu> {
  final _controller = WidgetStatesController();
  final _menuController = MenuController();
  final childNode = FocusNode(debugLabel: "Dropdown Inkwell");

  late final _actions = <Type, Action<Intent>>{
    DirectionalFocusIntent: _DirectionalFocusMenuAction(this),
    NextFocusIntent: _NextFocusMenuAction(this),
  };

  BuildContext get focusScopeContext => widget.firstItemFocusNode.context!;

  @override
  void dispose() {
    _controller.dispose();
    childNode.dispose();
    super.dispose();
  }

  void _handleTap() {
    if (!_menuController.isOpen) {
      _menuController.open();
      SchedulerBinding.instance.addPostFrameCallback((timeStamp) {
        FocusScope.of(context).setFirstFocus(
          widget.firstItemFocusNode.nearestScope!,
        );
        childNode.requestFocus();
      });
    } else {
      _menuController.close();
    }
  }

  @override
  Widget build(BuildContext context) {
    return MenuAnchor(
      controller: _menuController,
      childFocusNode: childNode,
      style: const MenuStyle(
        padding: WidgetStatePropertyAll(
            EdgeInsets.symmetric(vertical: 8, horizontal: 16)),
      ),
      builder: (context, controller, child) {
        return Actions(
          actions: _actions,
          child: InkWell(
            focusNode: childNode,
            statesController: _controller,
            borderRadius: const BorderRadius.vertical(top: Radius.circular(8)),
            onTap: _handleTap,
            child: ListenableBuilder(
              listenable: _controller,
              builder: (context, _) {
                return InputDecorator(
                  decoration:
                      const InputDecoration(border: UnderlineInputBorder()),
                  isFocused: _controller.value.contains(WidgetState.focused),
                  isHovering: _controller.value.contains(WidgetState.hovered),
                  child: Row(
                    children: [
                      Expanded(
                        child: DefaultTextStyle(
                          style:
                              Theme.of(context).dropdownMenuTheme.textStyle ??
                                  Theme.of(context).textTheme.bodyLarge!,
                          child: child!,
                        ),
                      ),
                      if (widget.trailingIcon != null) widget.trailingIcon!,
                    ],
                  ),
                );
              },
            ),
          ),
        );
      },
      menuChildren: widget.menuItems,
      child: widget.child,
    );
  }
}

class _DirectionalFocusMenuAction extends DirectionalFocusAction {
  _DirectionalFocusMenuAction(this.state);
  final _DropdownMenuState state;

  @override
  bool invoke(DirectionalFocusIntent intent) {
    if (intent.direction == TraversalDirection.up &&
        state.childNode.hasFocus &&
        state.widget.firstItemFocusNode.context?.mounted == true) {
      final FocusTraversalPolicy policy =
          FocusTraversalGroup.maybeOf(state.focusScopeContext) ??
              ReadingOrderTraversalPolicy();
      final lastFocus = policy.findLastFocus(
        state.widget.firstItemFocusNode,
        ignoreCurrentFocus: true,
      );
      lastFocus.requestFocus();
    } else {
      super.invoke(intent);
    }
    return true;
  }
}

class _NextFocusMenuAction extends NextFocusAction {
  _NextFocusMenuAction(this.state);
  final _DropdownMenuState state;

  @override
  bool invoke(NextFocusIntent intent) {
    if (state.childNode.hasFocus && state._menuController.isOpen) {
      state.widget.firstItemFocusNode.requestFocus();
      return true;
    }
    return super.invoke(intent);
  }
}
```

**Comment by PvtPuddles** [comment]
Your workaround works beautifully on the stable branch, all issues addressed.

I am eager for the remaining fixes to make their way through the system and I look forward to updates.

Thank you.

**Comment by davidhicks980** [comment]
Glad it helped! Yeah, hopefully the coming changes will make this workaround unnecessary -- we'll see soon.

**Comment by dkwingsmt** [comment]
Currently it works correctly on all platforms but Web, where down arrows doesn't produce any effects. I did some investigations and found out that it's because on other platforms, the down arrow is processed by the `Shortcuts` in `WidgetsApp`, where both `_defaultShortcuts` and `_defaultAppleOsShortcuts` dispatch a `DirectionalFocusIntent` for arrow keys. On Web, `_defaultWebShortcuts` dispatches `ScrollIntent` for arrow keys.

In short, it's the `DirectionalFocusIntent` that matters, no matter which widget dispatches it, and currently their dispatching has nothing to do with `MenuAnchor`s (although their processing does.)

**Comment by davidhicks980** [comment]
I can migrate the changes from RawMenuAnchor to fix this at some point. Not sure if @dkwingsmt is already working on it.

**Comment by Piinks** [comment]
> I can migrate the changes from RawMenuAnchor to fix this at some point. Not sure if @dkwingsmt is already working on it.

(Triage): Hey @davidhicks980, confirmed in triage here today - @dkwingsmt is not already working on this. All yours with gratitude! 🙏 💙

**Comment by davidhicks980** [comment]
Whoops @Piinks, I can't remember if I saw your comment. I'll try to get this done.

**Comment by davidhicks980** [comment]
I'm actually still working on this, but I'm chosing to wait until the material decoupling is finished. In the meantime, I've been working on https://github.com/davidhicks980/menu_utilities to figure out an optimal solution for keyboard traversal.
